### DISCUSSION TITLE
Model predictive control - Basics

### DISCUSSION THREAD

**Discussion by Cbapti** [yalmip/YALMIP#1357] [Problems using YALMIP] [answered]
Hello,

I have a question about the model predictive control - Basics Tutorial: https://yalmip.github.io/example/standardmpc/

The first parts were working fine for me, until the section **Improving simulation performance** comes up. 

After I implemented the suggested code:

```
ops = sdpsettings('verbose',2);

controller = optimize(constraints, objective,ops,x0,u{1}); 

x = [3;1];

for i = 1:5
 uk = controller{x};
 x = A*x + B*uk;
end
```

and start to run the script again, ill get an error message that says: 

_Warning: No suitable solver
Brace indexing is not supported for variables of this type.

Error in MPC_Basics (line 52)
 uk = controller{x};_

After I saw that i dont have a suitable solver i installed Mosek but that doesnt help. 

I use Matlab R2023b with the Optimization toolbox and Mosek installation.

Here is my full code:

```
A = [2 -1; 1 0.2];

B = [1;0];

nx = 2; % Num of States
nu = 1; % Num of inputs

% MPC data

Q = eye(2);
R = 2;
N = 7;

% Initial State

x0 = sdpvar(2,1); % Als optimierungsvariable definieren

u = sdpvar(repmat(nu,1,N),repmat(1,1,N)); % repmat erstellt eine matrix mit dem Wert, dass das erste Argument hat, mit der Form zweites Argument X drittes Argument

constraints = [];
objective = 0;

x = x0;

for k = 1:N
    x = A*x + B*u{k};

    objective = objective + norm(Q*x,1) + norm(R*u{k},1); % Formulierung der Zielfunktion die minimiert werden soll

    constraints = [constraints, -1 <= u{k} <= 1, -5 <= x <=5]; % Formulieren der Constraints
end

ops = sdpsettings('verbose',2);

controller = optimize(constraints, objective,ops,x0,u{1}); % Vorteil der Variante das erst hier x0 festegelegt wird

x = [3;1];

for i = 1:5
 uk = controller{x};
 x = A*x + B*uk;
end


for k = 1:N
    disp(value(u{k}));
end
```

Thanks in advance

**Comment by johanlofberg** [comment, marked as answer]
It's 

`controller = optimizer(...`

and 

Also, modern syntax of optimizer allows you to call it with values as any other function, i.e. `uk = controller(x); `which is more natural

**Reply by Cbapti** [reply]
Thank you!

**Comment by styw1234** [comment]
The code can be revised as 
`
%% Initial
clc
clear
close all
yalmip('clear')
rng(0)
%% Program
A = [2 -1; 1 0.2];

B = [1;0];

nx = 2; % Num of States
nu = 1; % Num of inputs

% MPC data

Q = eye(2);
R = 2;
N = 7;

% Initial State

x0 = sdpvar(2,1); % Als optimierungsvariable definieren

u = sdpvar(repmat(nu,1,N),repmat(1,1,N)); % repmat erstellt eine matrix mit dem Wert, dass das erste Argument hat, mit der Form zweites Argument X drittes Argument

constraints = [];
objective = 0;

x = x0;

for k = 1:N
    x = A*x + B*u{k};

    objective = objective + norm(Q*x,1) + norm(R*u{k},1); % Formulierung der Zielfunktion die minimiert werden soll

    constraints = [constraints, -1 <= u{k} <= 1, -5 <= x <=5]; % Formulieren der Constraints
end

ops = sdpsettings('verbose',2);

controller = optimize(constraints, objective); % Vorteil der Variante das erst hier x0 festegelegt wird

`
